### ISSUE TITLE
RLS does a better job of finding documentation and definitions

### ISSUE THREAD

**Issue by jacg** [rust-analyzer/rust-analyzer#4818]
Caveat: these are observations made in Emacs on NixOs with rust-src installed, I don't have the means to check in VSCode at present.

## Intra-crate
For example https://github.com/rust-analyzer/rust-analyzer/blob/ba821afa248153a58eede29048d2b21d30a2ea55/crates/ra_syntax/src/validation.rs#L3

- Placing the point on `block` gives

  + RLS: `Logic for validating block expressions i.e. ast::BlockExpr.` (i.e. the documentation of the `block` module)
  + RA: nothing

- Activating goto definition on `block` gives:

  + RLS: goes to `./validation/block.rs`
  + RA: goes to `./validation/block.rs` (i.e. the same as RLS)

## Extra-crate
For example https://github.com/rust-analyzer/rust-analyzer/blob/ba821afa248153a58eede29048d2b21d30a2ea55/crates/ra_syntax/src/validation.rs#L13

- Placing the point on `std`, `convert` or `TypeFrom` gives:

  + RLS: the documentation for The Rust Standard Library, `convert` or `TypeFrom`
  + RA: nothing

- Activating goto definition on `std`, `convert` or `TypeFrom` gives:

  + RLS: goes to the relevant source file in my `/nix/store`
  + RA:
    - for `std` goes to https://github.com/rust-analyzer/rust-analyzer/blob/ba821afa248153a58eede29048d2b21d30a2ea55/crates/ra_project_model/src/sysroot.rs#L44

     - for `convert` goes to https://github.com/rust-analyzer/rust-analyzer/blob/ba821afa248153a58eede29048d2b21d30a2ea55/crates/ra_hir_ty/src/infer.rs#L99

     - for `TypeFrom`: does nothing.

Note, even if the source isn't in `/nix/store` but in `~/.cargo`, RLS succeeds, but RA fails, for example on `EE` or `ZeroChars` at 

https://github.com/rust-analyzer/rust-analyzer/blob/ba821afa248153a58eede29048d2b21d30a2ea55/crates/ra_syntax/src/validation.rs#L20

## Summary

In a number of situations where RLS finds documentation and definitions, RA fails to find anything, or goes to an unrelated definition of the *unqualified* name (c.f. `std` and `convert` above).

**Comment by flodiebold** [comment]
Go to def for `std`, `convert` and `TryFrom` works for me. This might be a setup problem? Do you get completions for std?

You're right that we don't display documentation for modules on hover.

`EE::ZeroChars` is a local import, which isn't supported yet (#1165).

**Comment by jacg** [comment]
> Do you get completions for std?

If I type `use std::` (in `validation.rs`)

+ RLS suggests `convert`, `string`, `str`, `char`, `ascii`, `alloc` and many, many more (*all* annotated with paths in `/nix/store`). If I continue to type, say `col`, it narrows it down to `collections`.
+ RA suggests exactly two names: `self` and `super` annotated with `(Keyword)`.

> This might be a setup problem?

Entirely possible. I'm using Spacemacs with the Rust layer. Here is the sum total of my Spacemacs Rust configuration

```elisp
(rust :variables
      rust-backend 'lsp
      lsp-rust-server 'rust-analyzer
      lsp-ui-sideline-actions-kind-regex ".*" ;; rust-analyzer assists
      )
```
I'm in a nix-shell (impure, started with direnv) defined with

```nix
let pkgs = import <nixpkgs> {
      overlays = map (uri: import (fetchTarball uri)) [
        https://github.com/mozilla/nixpkgs-mozilla/archive/master.tar.gz
      ];
    };
in
  with pkgs;
  stdenv.mkDerivation {
    name = "rust-wasm";
    buildInputs = [
      (latest.rustChannels.nightly.rust.override {
        extensions = ["rust-src"];
      })
    ];
  }
```

Not sure how to hunt for configuration problems here.

**Comment by jacg** [comment]
> You're right that we don't display documentation for modules on hover.

Is that by choice, or has nobody got around to it, or is there some difficulty in implementing it?

**Comment by kjeremy** [comment]
It's because of #2148

The first step was completed recently but inner doc comments are not attached yet.

**Comment by bjorn3** [comment]
Did you start spacemacs from that nix-shell?

**Comment by jacg** [comment]
@bjorn3 I have direnv inside my Emacs, so it loads any relevant nix-shell automatically. I'm pretty confident that it's finding the rust installation, as I can compile from within Emacs, and RLS is finding Rust sources in `/nix/store` which are only present when that nix-shell is active, and so on.

**Comment by jacg** [comment]
@kjeremy 

>  PRs welcome.

How newcomer-friendly is this likely to be?

**Comment by flodiebold** [comment]
It looks like rust-analyzer can't find the sysroot. Can you run `rustc --print sysroot` from within Emacs (e.g. using `shell-command`)?

**Comment by jacg** [comment]
Both the Emacs in which I'm running RLS and the one in which I'm running RA agree that `rustc --print sysroot` is `/nix/store/q77rj11qh9i9v7cafyx16b1xygr8nk62-rust-1.46.0-nightly-2020-06-08-fd4b177aa`

**Comment by flodiebold** [comment]
So does `/nix/store/q77rj11qh9i9v7cafyx16b1xygr8nk62-rust-1.46.0-nightly-2020-06-08-fd4b177aa/lib/rustlib/src/` exist and contain the std source?

**Comment by jacg** [comment]
... hmm, but my RLS comes from the nix-shell, and my RA is self-compiled ...

**Comment by jacg** [comment]
> So does `/nix/store/q77rj11qh9i9v7cafyx16b1xygr8nk62-rust-1.46.0-nightly-2020-06-08-fd4b177aa/lib/rustlib/src/` exist and contain the std source?

Yes, and that is exactly where RLS finds it. But it is also where RLS itself lives, while RA does not:

```bash

$ which rls
/nix/store/q77rj11qh9i9v7cafyx16b1xygr8nk62-rust-1.46.0-nightly-2020-06-08-fd4b177aa/bin/rls

$ which rust-analyzer
/home/jacek/.local/bin/rust-analyzer
```

**Comment by flodiebold** [comment]
Hmm but that shouldn't be a problem, if rust-analyzer finds the files. Do you happen to have the `RUST_SRC_PATH` env variable set (you shouldn't)? If not, do you see any error message in the logs from rust-analyzer? (You may need to run it with `RA_LOG=error`, not sure.)

**Comment by jacg** [comment]
`RUST_SRC_PATH` is not set.

**TL;DR:** I can't find any error messages in the logs, here are the details ...

With `RA_LOG` unset, nothing much appears in any of the `*lsp-log*`, `*rust-analyzer*` or `*rust-analyzer::stderr*` buffers.

When setting `RA_LOG=error`, nothing changes in the first two, but `*rust-analyzer::stderr` fills with an almost infinite amount of stuff, that is to say merely opening `validate.rs` generates almost 750k (7.5 x 10 ^ 5) lines. I can't find any interesting errors there, but that's a pretty big haystack.

If I then perform a goto definition on the `std` on line 13 (as in my original comment) it jumps almost immediately to the same incorrect location as reported above, and the CPU fan spins for quite some time as another 180k lines get fed into `*rust-analyzer::stderr*`. Near the beginning of those, I find

    [INFO  rust_analyzer::main_loop] loop turn = Request(Request { id: RequestId(U64(45)), method: "textDocument/definition", params: Object({"position": Object({"character": Number(5), "line": Number(12)}), "textDocument": Object({"uri": String("file:///home/jacek/rust/rust-analyzer/crates/ra_syntax/src/validation.rs")})}) })

which, accounting for LSP's zero-based counting, seems to fit the bill.

After 2913 lines of mostly

    [INFO  salsa::derived::slot] ParseQuery(FileId(2678)): invoked at R11853
    [INFO  salsa::derived::slot] ParseQuery(FileId(2678)): executing query
    [INFO  salsa::derived::slot] FileSymbolsQuery(FileId(3295)): invoked at R11853
    [INFO  salsa::derived::slot] FileSymbolsQuery(FileId(3295)): executing query
    [INFO  salsa::derived::slot] ParseQuery(FileId(3295)): invoked at R11853
    [INFO  salsa::derived::slot] ParseQuery(FileId(3295)): executing query
    [INFO  salsa::derived::slot] FileSymbolsQuery(FileId(3297)): invoked at R11853
    [INFO  salsa::derived::slot] FileSymbolsQuery(FileId(3297)): executing query
    [INFO  salsa::derived::slot] ParseQuery(FileId(3297)): invoked at R11853

we eventually get to

    [INFO  rust_analyzer::main_loop] loop turn = Response { id: RequestId(U64(45)), error: None }

That's the same `RequestId` as the original request (45), and I guess that `error: None` means 'there was no error' as opposed to 'There was an error: something returned `None`'.

In short, I cannot find any errors.

**Comment by flodiebold** [comment]
Hm. I wouldn't expect to see `INFO` logs if `RA_LOG` is set to `error`. It should only show error logs (and there shouldn't be a lot of them).

**Comment by flodiebold** [comment]
I'll close this since we've got separate bugs for all the actual issues, and the rest seems like a setup problem.
